### PR TITLE
Loop over python versions in template project

### LINKED ISSUE #12: Can we specify a list of python versions and loop through them?
The combination of https://github.com/UCL-ARC/python-template/blob/49898d88e6f46e0e214b0d4eecee2d07c19cad82/cookiecutter.json#L9 and https://github.com/UCL-ARC/python-template/blob/49898d88e6f46e0e214b0d4eecee2d07c19cad82/%7B%7Bcookiecutter.project_slug%7D%7D/pyproject.toml#L109-L129 doesn't work at the moment. What I think would be good would be specifying either a list of python versions to support or a min and/or max version and do something clever.

### PR DESCRIPTION
**Pull request by paddyroddy**
Fixes #12. See also #13 for why I've removed `setuptools-scm`.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,5 +1,38 @@
 repos:
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: v0.0.255
+    hooks:
+      - id: ruff
+        args: ["--config", "pyproject.toml"]
```
Comment (paddyroddy) [inline .pre-commit-config.yaml:6]: This makes sure it uses the top one and not the one in the template. See charliermarsh/ruff#1812.

Comment (p-j-smith) [reply .pre-commit-config.yaml:6]: that's handy to know!

File: {{cookiecutter.project_slug}}/pyproject.toml [hunk 2]
```diff
@@ -110,8 +107,11 @@ trailing_comma_inline_array = true
 legacy_tox_ini = """
     [gh-actions]
     python =
-    {%- for version in python_versions -%}
-        version: py{{ version | replace: '.', ''}
+    {% for python_version in range(
+        cookiecutter.min_python_version | replace('3.', '') | int,
+        cookiecutter.max_python_version | replace('3.', '') | int
+    )%}
+        3.{{ python_version }}: py3{{ python_version }}
```
Comment (p-j-smith) [inline {{cookiecutter.project_slug}}/pyproject.toml:114]: at the minute this formats the list of versions like this:
```
python =

3.8: py38

3.9: py39

3.10: py310

``` 
I think it should format without newlines if you move `3.{{ python_version }}: py3{{ python_version }}` to the same line as the closing brace in the for loop

File: {{cookiecutter.project_slug}}/pyproject.toml [hunk 3]
```diff
@@ -122,8 +122,11 @@ legacy_tox_ini = """
...
     [tox]
-    env_list = 
-    {%- for version in python_versions -%}
-        py{{ version | replace: '.', ''}
+    env_list =
+    {% for python_version in range(
+        cookiecutter.min_python_version | replace('3.', '') | int,
+        cookiecutter.max_python_version | replace('3.', '') | int
+    )%}
+        py3{{ python_version }}
```
Comment (p-j-smith) [inline {{cookiecutter.project_slug}}/pyproject.toml:130]: same as above about the formatting

**Review by p-j-smith — CHANGES_REQUESTED** [review]
very nice! just a  small couple of changes for the tox config formatting

File: {{cookiecutter.project_slug}}/pyproject.toml (hunk 2 above)
Comment (paddyroddy) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: Ah yeah didn't think to look at that. Oops

Comment (paddyroddy) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: It doesn't look like that on mine! But will see if I can remove the whitespace

Comment (paddyroddy) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: Looks like this for me
```
[tool.tox]
legacy_tox_ini = """
    [gh-actions]
    python =
    
        3.8: py38
    
        3.9: py39
    
        3.10: py310
    

    [testenv]
    commands =
        pytest --cov
    deps =
        pytest
        pytest-cov

    [tox]
    env_list =
    
        py38
    
        py39
    
        py310
    
"""
```

Comment (p-j-smith) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: for some reason you need to commit the change before you can run cookiecutter on your local folder and see the changes. I guess it still relies on git history when running locally?

Comment (p-j-smith) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: after commiting,  that section looks like this for me:
```
    [gh-actions]
    python =
     3.8: py38
     3.9: py39
     3.10: py310
    

    [testenv]
```
So maybe also remove the newline before `[testenv]` in the template

Comment (p-j-smith) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: sorry just saw you've already fixed it!

**Review by p-j-smith — APPROVED** [review]

File: {{cookiecutter.project_slug}}/pyproject.toml (hunk 2 above)
Comment (paddyroddy) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: The key is using i.e. `{%-` rather than `{%` to remove whitespace before

Comment (p-j-smith) [reply {{cookiecutter.project_slug}}/pyproject.toml:114]: ah that's good to know, thanks!

**Comment by paddyroddy** [comment]
Yeah I'm still working on this as spotted a couple of issues so wait for now

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -7,7 +7,11 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version:
+          - "3.8"
+          - "3.9"
+          - "3.10"
+          - "3.11"
```
Comment (paddyroddy) [inline .github/workflows/test.yml:14]: This is just to be consistent with the one inside the template

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -1,5 +1,38 @@
...
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v1.1.1
+    hooks:
+      - id: mypy
   - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: v3.0.0-alpha.4
+    rev: v3.0.0-alpha.6
     hooks:
       - id: prettier
+        exclude: "{{cookiecutter.project_slug}}/.github/workflows/tests.yml"
```
Comment (paddyroddy) [inline .pre-commit-config.yaml:23]: Prettier hates syntax which works

File: {{cookiecutter.project_slug}}/.github/workflows/tests.yml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version:
+        {%- for python_version in range(
+          cookiecutter.min_python_version | replace('3.', '') | int,
+          cookiecutter.max_python_version | replace('3.', '') | int
+        ) %}
+          - "3.{{ python_version }}"
+        {%- endfor %}
```
Comment (paddyroddy) [inline {{cookiecutter.project_slug}}/.github/workflows/tests.yml]: I could get this syntax to work, but not the `[]` one. Missed this first time around.

**Comment by paddyroddy** [comment]
@p-j-smith can you have another look? Also run `cookieninja https://github.com/UCL-ARC/python-template --checkout fix-python-versions` and double check it all looks right

**Comment by p-j-smith** [comment]
> can you have another look? Also run cookieninja https://github.com/UCL-ARC/python-template --checkout fix-python-versions and double check it all looks right

yep, it works! I installed it in a clean environment and ran tox, tests pass. But currently the tox config is:
```.toml
[tool.tox]
legacy_tox_ini = """
    [gh-actions]
    python =
        3.8: py38
        3.9: py39
        3.10: py310

    [testenv]
    commands =
        pytest --cov
    deps =
        pytest
        pytest-cov

    [tox]
    env_list =
        py38
        py39
        py310
"""
```
So it's missing testing for 3.11. And same for the `tests` workflow

**Comment by paddyroddy** [comment]
> > can you have another look? Also run cookieninja https://github.com/UCL-ARC/python-template --checkout fix-python-versions and double check it all looks right
> 
> yep, it works! I installed it in a clean environment and ran tox, tests pass. But currently the tox config is:
> 
> ```toml
> [tool.tox]
> legacy_tox_ini = """
>     [gh-actions]
>     python =
>         3.8: py38
>         3.9: py39
>         3.10: py310
> 
>     [testenv]
>     commands =
>         pytest --cov
>     deps =
>         pytest
>         pytest-cov
> 
>     [tox]
>     env_list =
>         py38
>         py39
>         py310
> """
> ```
> 
> So it's missing testing for 3.11. And same for the `tests` workflow

So I had the same issue. Any idea what might be causing it?

**Comment by paddyroddy** [comment]
Ah it will be `range`! Stupid me. Hang on

**Comment by paddyroddy** [comment]
Fixed!

**Review by p-j-smith — APPROVED** [review]

**Comment by p-j-smith** [comment]
> Fixed!

awesome!

**Pull request merged**